CANTroller2/src/devices.h

```diff
@@ -158,125 +200,75 @@ enum class TransducerDirection : uint8_t {REV, FWD}; // possible dir values. REV
+            ret = _b_offset + (_m_factor * arg_val_f);
+        } else if (arg_val_f) { // NOTE: isn't 0.0 a valid value tho?
+            if (dir == TransducerDirection::REV) {
+                ret = min_f + (max_f - (_b_offset + (_m_factor / arg_val_f)));
+            }
+            ret = _b_offset + (_m_factor / arg_val_f);
+        } else {
+            printf ("Error: unit conversion refused to divide by zero\n");
+            // NOTE: hmmmm, couldn't -1 be a valid value in some caes?
+        }
+        return static_cast<HUMAN_T>(ret);
+    }
+
+    virtual NATIVE_T to_native(HUMAN_T arg_val_human) {
+        float arg_val_f = static_cast<float>(arg_val_human); // convert everything to floats so we don't introduce rounding errors
+        float min_f = static_cast<float>(human.get_min());
+        float max_f = static_cast<float>(human.get_max());
+        float ret = -1;
+        if (dir == TransducerDirection::REV) {
+            arg_val_f = min_f + (max_f - arg_val_f);
+        }
+        if (_invert && (arg_val_f - _b_offset)) {
+            ret = _m_factor / (arg_val_f - _b_offset);
+        } else if (!_invert && _m_factor) {
+            ret = (arg_val_f - _b_offset) / _m_factor;
+        } else {
+            printf ("Error: unit conversion refused to divide by zero\n");
+            // NOTE: hmmmm, couldn't -1 be a valid value in some caes?
+        }
+        return static_cast<NATIVE_T>(ret);
+    }
 
     // NOTE: do we really need two values? or should this just be a single value and get converted wherever needed?
     // To hold val/min/max display values in display units (like V, mph, etc.)
     Param<HUMAN_T> human;
     Param<NATIVE_T> native;
-
-    TransducerDirection dir = TransducerDirection::FWD; // Belongs in a child class for devices. For the case a lower val causes a higher real-life effect, 
-
-    virtual HUMAN_T from_native(NATIVE_T arg_val_native) = 0;
-    virtual NATIVE_T to_native(HUMAN_T arg_val_native) = 0;
-
   public:
     Transducer(uint8_t arg_pin) : Device(arg_pin) {}
     Transducer() = delete;
 
     // NOTE: do we want to be able to set min/max separately?
-    // NOTE: do we want to be able to have limits maintained internally (so not references)?
-    // NOTE: since we really only have one value and two representations, should these limit calls change both limits?
+    // NOTE: changing the limits of one type doesn't affect the limits of the other type, should it?
+    // NOTE: changing the limits of either type will cause the 
     void set_native_limits(Param<NATIVE_T> &arg_min, Param<NATIVE_T> &arg_max) {
         if (arg_min.get() > arg_max.get()) {
             dir = TransducerDirection::REV;
             native.set_limits(arg_max, arg_min);
-            // NOTE: should change the human val as well if constrained
         }
         else {
             dir = TransducerDirection::FWD;
             native.set_limits(arg_min, arg_max);
         }
+        human.set(from_native(native.get()));
```

Comment:
I tried this and ran into an issue with converting the limits back and forth. I think it could be solved (although I think we'd keep human and not native, since we do all the rest of the logic in the human-readable value), but it would be a substantial re-write, so I'll tackle it in a separate PR.